### PR TITLE
Added support of ipv6 address (boolean "use_v6" parameter in TelegramClient constructor)

### PR DESCRIPTION
**Pull request by Man197**
Hello! Well, I did that:) 
With this mod the library can work with my [PySocks fork](https://github.com/Man197/PySocks) to use IPv6 address.

### CODE REVIEW

File: telethon/telegram_bare_client.py [hunk 1]
```diff
@@ -69,6 +69,7 @@ class TelegramBareClient:
 
     def __init__(self, session, api_id, api_hash,
                  connection_mode=ConnectionMode.TCP_FULL,
+                 use_v6=False,
```
Comment (Lonami) [inline telethon/telegram_bare_client.py]: `v6` sounds like "version six". Version six of what? Maybe something more explicit like `use_ipv6` would make more sense. I'll change this.

File: telethon/tl/session.py [hunk 2]
```diff
@@ -109,15 +112,15 @@ def list_sessions():
...
 
     @staticmethod
-    def try_load_or_create_new(session_user_id):
+    def try_load_or_create_new(session_user_id, use_v6=False):
         """Loads a saved session_user_id.session or creates a new one.
            If session_user_id=None, later .save()'s will have no effect.
         """
         if session_user_id is None:
-            return Session(None)
+            return Session(None, use_v6)
```
Comment (Lonami) [inline telethon/tl/session.py]: I'm not too sure about this, having to pass `use_v6` around everywhere. I will change this to have `None` IP address, and if this is the case, get the right IP from somewhere else.

File: telethon/tl/session.py [hunk 3]
```diff
@@ -64,7 +64,10 @@ def __init__(self, session_user_id):
         self._last_msg_id = 0  # Long
 
         # These values will be saved
-        self.server_address = '91.108.56.165'
+        if use_v6:
+            self.server_address = '[2001:67c:4e8:f002::a]'
```
Comment (Lonami) [inline telethon/tl/session.py]: So maybe the problem was the address wasn't enclosed with square braces? Why is this?

**Comment by Lonami** [comment]
Thank you! However I've noticed you don't use the `use_v6` information when a reconnection occurs, which is likely. See this method:
https://github.com/LonamiWebs/Telethon/blob/5a57a8a498b1461d8aa041cc9adb7ec71eb88a33/telethon/telegram_bare_client.py#L290

File: telethon/telegram_bare_client.py (hunk 1 above)
Comment (Man197) [reply telethon/telegram_bare_client.py]: Yea, I also think that it will be better.

File: telethon/tl/session.py (hunk 3 above)
Comment (Man197) [reply telethon/tl/session.py]: AFAIK, ipv6 address must be enclosed in square brackets if it used with port or within URL. But I'm not exactly sure, that these brackets are needed also there.

**Comment by Man197** [comment]
I missed this place in the code. Do you have any suggestions on how best to throw a flag from the class constructor into this function?

**Comment by Lonami** [comment]
I've made some changes to your code so it's a bit cleaner. Please review them and let me know if this is enough to help you do that final change I requested.

**Comment by Man197** [comment]
I see that ipv6 parameter is used only in the same function: https://github.com/LonamiWebs/Telethon/blob/5a57a8a498b1461d8aa041cc9adb7ec71eb88a33/telethon/telegram_bare_client.py#L311
So, maybe the best way will be to store use_ipv6 in class field and check it within this function?

**Comment by Lonami** [comment]
Yeah, best option in my opinion is to have `TelegramBareClient._use_ipv6` so we can use `self._use_ipv6` instead passing `ipv6=` as a parameter. We also need to think about this case: what if an user is using IPv4 but decides to use IPv6 at some point? It can happen, overall if you're testing.

**Comment by Man197** [comment]
Anyway, the originally selected IP is saved in the session, so I think we can make the field private and fill it only when creating an instance of the class (or restoring the session from the file).

File: telethon/telegram_bare_client.py [hunk 4]
```diff
@@ -300,15 +313,15 @@ def _get_dc(self, dc_id, ipv6=False, cdn=False):
 
             return next(
                 dc for dc in TelegramBareClient._config.dc_options
-                if dc.id == dc_id and bool(dc.ipv6) == ipv6 and bool(dc.cdn) == cdn
+                if dc.id == dc_id and bool(dc.ipv6) == self._use_ipv6 and bool(dc.cdn) == cdn
```
Comment (Lonami) [inline telethon/telegram_bare_client.py]: Here again, I noticed you use `'[2001:67c:4e8:f002::a]'` as the IPv6 string (notice the square braces `'[]'`). Why are these needed? Telegram's `dc.ipv6` don't have these braces. Will it still work?

Comment (Man197) [reply telethon/telegram_bare_client.py]: OK, I will test it today/tomorrow and let you know.

Comment (Lonami) [reply telethon/telegram_bare_client.py]: Easiest way is to manually set a different datacenter on the session file so when you connect the library will need to change to your closest one.

Comment (Man197) [reply telethon/telegram_bare_client.py]: Well, as I've just tested, in case of ipv6 connection, the server address MUST be in `'[]'` and the port MUST be 80, or server/proxy will return an Bad Gateway or Timeout error.

**Comment by Lonami** [comment]
I'm testing this pull request now and while I can't connect I found [this](https://docs.python.org/3/library/socket.html?highlight=socket#socket.has_ipv6):
> socket.**has_ipv6**
> This constant contains a boolean value which indicates if IPv6 is supported on this platform.

So maybe we should check `if use_ipv6 and not socket.has_ipv6: raise`.

**Comment by Lonami** [comment]
Even better:
```python
>>> socket.has_ipv6
True
>>> socket.getaddrinfo("example.org", 80, proto=socket.IPPROTO_TCP)[1]
(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('2606:2800:220:1:248:1893:25c8:1946', 80, 0, 0))
>>> a = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
>>> a.connect(('2606:2800:220:1:248:1893:25c8:1946', 80, 0, 0))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 101] Network is unreachable
>>>
>>> socket.getaddrinfo("example.org", 80, proto=socket.IPPROTO_TCP)[0]
(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('93.184.216.34', 80))
>>> a = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
>>> a.connect(('93.184.216.34', 80))
>>> a.close()
```

**Comment by Man197** [comment]
AFAIK, even if ipv6 is supported on the target platform, the machine may not connect via ipv6 without having an ipv6 address. It depends on the ISP. An alternative way is to use the ipv6 proxy, for which all these changes were started.
`socket.has_ipv6` does not provide reliable information, and it seems to me that there is no difference that it will cause an exception - Telethon or socket module. Therefore, I think check `socket.has_ipv6` is superfluous.

**Comment by Lonami** [comment]
Alright, thanks for the info.

**Comment by Lonami** [comment]
Despite this though, `ifconfig` does show an IPv6 address on the network adapter I'm using so I think it's safe to assume that I do have IPv6 (also, I can create an IPv6 socket without trouble).

**Comment by Man197** [comment]
Your network adapter can have an ipv6 address, so `socket.has_ipv6` also returns True. But does your ISP network have ipv6 addresses? Most likely, all use of ipv6 ends on your network adapter :)

**Comment by Lonami** [comment]
Thanks, I will add support for switching between IPv4 to IPv6 and vice versa if a previous connection was made now.

**Comment by Man197** [comment]
Thank you so much! Many people have already asked me to do pull request, finally they will be happy :)

**Comment by Lonami** [comment]
One last note, Telegram as I said returns this for IPv6 addresses:
```python
dc_options=[
    DcOption(
        ipv6=False,
        media_only=False,
        tcpo_only=False,
        cdn=False,
        static=False,
        id=1,
        ip_address='149.154.175.50',
        port=443
    ),
    DcOption(
        ipv6=True,
        media_only=False,
        tcpo_only=False,
        cdn=False,
        static=False,
        id=1,
        ip_address='2001:0b28:f23d:f001:0000:0000:0000:000a',
        port=443
    ),
    ...
],
```

So I'll add the square braces to surround the IP address if IPv6 is enabled, as you told me it needs to be.
